daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ExpressionEvaluatingUnparsers.scala

```diff
@@ -77,30 +76,52 @@ final class SetVariableUnparser(
 
 }
 
+final class NewVariableInstanceSuspendableExpression(
+  override val expr: CompiledExpression[AnyRef],
+  override val rd: VariableRuntimeData)
+  extends SuspendableExpression {
+
+  override protected def processExpressionResult(ustate: UState, v: DataValuePrimitive): Unit = {
+    ustate.variableMap.newVariableInstance(rd, v)
+  }
+
+  override protected def maybeKnownLengthInBits(ustate: UState) = MaybeULong(0)
+}
+
 // When implemented this almost certainly wants to be a combinator
 // Not two separate unparsers.
-class NewVariableInstanceStartUnparser(override val context: RuntimeData)
+class NewVariableInstanceStartUnparser(override val context: VariableRuntimeData)
   extends PrimUnparserNoData {
 
   override lazy val runtimeDependencies = Vector()
 
   override lazy val childProcessors = Vector()
 
+  def suspendableExpression = {
+    Assert.invariant(context.maybeDefaultValueExpr.isDefined)
+    new NewVariableInstanceSuspendableExpression(context.maybeDefaultValueExpr.get, context)
+  }
+
   override def unparse(state: UState) = {
-    val vrd = context.asInstanceOf[VariableRuntimeData]
-    state.newVariableInstance(vrd)
+    if (context.maybeDefaultValueExpr.isDefined) {
+      suspendableExpression.run(state)
+    }
+    else {
+      state.variableMap.newVariableInstance(context)
+    }
   }
 }
 
-class NewVariableInstanceEndUnparser(override val context: RuntimeData)
+class NewVariableInstanceEndUnparser(override val context: VariableRuntimeData)
   extends PrimUnparserNoData {
 
   override lazy val runtimeDependencies = Vector()
 
   override lazy val childProcessors = Vector()
 
   override def unparse(state: UState) = {
-    state.removeVariableInstance(context.asInstanceOf[VariableRuntimeData])
+    if (state.variableMap.find(context.globalQName).isDefined)
```

Comment:
In what case coudl the variable not exist?

Is what's going on here is the newVariableInstance suspends and so hasn't been created yet. We continue unparsing but because that instance wasn't created so we have nothing to create so nothing to delete?

But, say we had two nested new variable instances. Say the outer one doesn't suspend and the inner one does. In that case, the outer NewVariableInstanceStartUnparser adds a new variable instance just fine. Then we call NewVariableInstanceStartUnparser for the inner variable instance. That inner NVI suspends and so doesn't create the variable isntance until much later. Unparse will continue and we'll call NewVariableInstanceEndUnparser for the inner one. Note that we still haven't created the outer new variable instance. But we do have a new variable instance from the outter NVI. So this find() will work, and we'll reomve a variable instance. But we'll remove the outer NVI, which this should only remove the inner NVI (which doesn't exist yet). So now variable access will likely fail because we've removed our instance.

Does that all sound like what would happen, or am I missing something subtle? If that is right, then it probably leads to my previous comment that we do need to always create an NVI without an value, and then suspensions can modify those values.

Reply:
I think SL is correct. This isn't right. We can't suspend creation of the variable instance, we have to suspend the calculation of its value sometimes, and we can't read it until that calculation is complete, but the instance has to be there so that the expressions can try to read and determine that they've blocked on a pending variable value.